django/db/models/functions/math.py

```diff
@@ -0,0 +1,164 @@
+import math
+import sys
+
+from django.db.models import DecimalField, FloatField, Func, Transform
+from django.db.models.functions import Cast
+
+
+class DecimalInputMixin:
+
+    def as_postgresql(self, compiler, connection):
+        # Cast FloatField to DecimalField as PostgreSQL doesn't support the
+        # following function signatures by default:
+        # - LOG(double precision, double precision)
+        # - MOD(double precision, double precision)
+        output_field = DecimalField(decimal_places=sys.float_info.dig, max_digits=1000)
+        clone = self.copy()
+        clone.set_source_expressions([
+            Cast(expression, output_field) if isinstance(expression.output_field, FloatField)
+            else expression for expression in self.get_source_expressions()
+        ])
+        return clone.as_sql(compiler, connection)
+
+
+class OutputFieldMixin:
+
+    def _resolve_output_field(self):
+        sources = self.get_source_expressions()
+        if any(isinstance(s.output_field, DecimalField) for s in sources):
+            return DecimalField()
+        else:
+            return FloatField()
+
+
+class Abs(Transform):
+    function = 'ABS'
+    lookup_name = 'abs'
+
+
+class ACos(OutputFieldMixin, Transform):
+    function = 'ACOS'
+    lookup_name = 'acos'
+
+
+class ASin(OutputFieldMixin, Transform):
+    function = 'ASIN'
+    lookup_name = 'asin'
+
+
+class ATan(OutputFieldMixin, Transform):
+    function = 'ATAN'
+    lookup_name = 'atan'
+
+
+class ATan2(OutputFieldMixin, Func):
+    function = 'ATAN2'
+    arity = 2
+
+
+class Ceil(Transform):
+    function = 'CEILING'
+    lookup_name = 'ceil'
+
+    def as_oracle(self, compiler, connection):
+        return super().as_sql(compiler, connection, function='CEIL')
+
+
+class Cos(OutputFieldMixin, Transform):
+    function = 'COS'
+    lookup_name = 'cos'
+
+
+class Cot(OutputFieldMixin, Transform):
+    function = 'COT'
+    lookup_name = 'cot'
+
+    def as_oracle(self, compiler, connection):
+        return super().as_sql(compiler, connection, template='(1 / TAN(%(expressions)s))')
```

Comment:
Following on from the previous comment, if we were to make the change, this should also be changed accordingly.